libeos-update-server/config.c

```diff
@@ -0,0 +1,213 @@
+
+/* Configuration file keys. */
+static const char *LOCAL_NETWORK_UPDATES_GROUP = "Local Network Updates";
+static const char *ADVERTISE_UPDATES_KEY = "AdvertiseUpdates";
+
+static const gchar *REPOSITORY_GROUP = "Repository ";  /* should be followed by an integer */
+static const gchar *PATH_KEY = "Path";
+static const gchar *REMOTE_NAME_KEY = "RemoteName";
+
+/**
+ * eus_repo_config_free:
+ * @config: (transfer full): an #EusRepoConfig
+ *
+ * Free the given @config, which must be non-%NULL.
+ *
+ * Since: UNRELEASED
+ */
+void
+eus_repo_config_free (EusRepoConfig *config)
+{
+  g_free (config->remote_name);
+  g_free (config->path);
+  g_free (config);
+}
+
+static EusRepoConfig *
+eus_repo_config_new_steal (guint  index,
+                           gchar *path,
+                           gchar *remote_name)
+{
+  g_autoptr(EusRepoConfig) config = NULL;
+
+  config = g_new0 (EusRepoConfig, 1);
+  config->index = index;
+  config->path = g_steal_pointer (&path);
+  config->remote_name = g_steal_pointer (&remote_name);
+
+  return g_steal_pointer (&config);
+}
+
+/**
+ * eus_read_config_file:
+ * @config_file_path: (nullable): path to the configuration file, or %NULL to
+ *    use the system search paths
+ * @advertise_updates: (out caller-allocates) (optional): return location for
+ *    the `AdvertiseUpdates=` parameter
+ * @repository_configs: (out callee-allocates) (transfer container)
+ *    (element-type EusRepoConfig) (optional): return location for the
+ *    `[Repository 0–9]` sections
+ * @error: return location for a #GError, or %NULL
+ *
+ * Find and load the `eos-update-server.conf` configuration file. If
+ * @config_file_path is non-%NULL, the file will be loaded from that path.
+ * Otherwise, it will be loaded from the system search paths as documented in
+ * [`eos-update-server.conf(5)`](man:eos-update-server.conf(5)).
+ *
+ * The configuration values loaded from the file will be returned in
+ * @advertise_updates and @repository_configs. See
+ * [`eos-update-server.conf(5)`](man:eos-update-server.conf(5)) for the
+ * semantics of the options.
+ *
+ * Each #EusRepoConfig element in the returned @repository_configs array
+ * contains the options from a single `[Repository 0–9]` section.
+ *
+ * Returns: %TRUE on success, %FALSE otherwise
+ * Since: UNRELEASED
+ */
+gboolean
+eus_read_config_file (const gchar  *config_file_path,
+                      gboolean     *advertise_updates,
+                      GPtrArray   **repository_configs,
+                      GError      **error)
+{
+  g_autoptr(GKeyFile) config = NULL;
+  g_autoptr(GError) local_error = NULL;
+  const gchar * const default_paths[] =
+    {
+      CONFIG_FILE_PATH,
+      LOCAL_CONFIG_FILE_PATH,
+      STATIC_CONFIG_FILE_PATH,
+      NULL
+    };
+  const gchar * const override_paths[] =
+    {
+      config_file_path,
+      NULL
+    };
+  g_auto(GStrv) groups = NULL;
+  gsize n_groups, i;
+  gboolean _advertise_updates;
```

Comment:
Cue in some language lawyer blathering about identifiers starting with an underline being reserved for C language/standard library.

Maybe remove the leading underline from the variable names and rename the function out parameters to something like `out_advertise_updates` or something?